### PR TITLE
TDL-15865: Add request timeouts

### PR DESCRIPTION
**Pull request by hpatel41**
# Description of change
TDL-15865: Add request timeouts

- Timeout for requests was already added, as this Tap is using SOAP and REST hence, added backoff for the requests.
- Updated the default timeout from 900 to 300 seconds.

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_exacttarget/client.py [hunk 1]
```diff
@@ -6,6 +6,9 @@
 
 LOGGER = singer.get_logger()
 
+# default timeout for requests
```
Comment (dbshah1212) [inline tap_exacttarget/client.py:9]:
```suggestion
# default request timeout
```

File: tap_exacttarget/client.py [hunk 2]
```diff
@@ -52,7 +55,15 @@ def get_auth_stub(config):
...
         params['useOAuth2Authentication'] = "False"
         auth_stub = FuelSDK.ET_Client(params=params)
-        transport = HttpAuthenticated(timeout=int(config.get('request_timeout', 900)))
+
+        # Set request timeout with config param `request_timeout`.
+        config_request_timeout = config.get('request_timeout')
+        if config_request_timeout and float(config_request_timeout):
+            request_timeout = float(config_request_timeout)
+        else:
+            request_timeout = REQUEST_TIMEOUT # If value is 0,"0","" or not passed then it set default to 300 seconds.
```
Comment (dbshah1212) [inline tap_exacttarget/client.py:64]:
```suggestion
            request_timeout = REQUEST_TIMEOUT # If value is 0, "0", "" or not passed then it sets default to 300 seconds.
```

File: tap_exacttarget/client.py [hunk 3]
```diff
@@ -73,7 +84,15 @@ def get_auth_stub(config):
...
         LOGGER.info("Authentication URL is: %s", params['authenticationurl'])
         auth_stub = FuelSDK.ET_Client(params=params)
-        transport = HttpAuthenticated(timeout=int(config.get('request_timeout', 900)))
+
+        # Set request timeout with config param `request_timeout`.
+        config_request_timeout = config.get('request_timeout')
+        if config_request_timeout and float(config_request_timeout):
+            request_timeout = float(config_request_timeout)
+        else:
+            request_timeout = REQUEST_TIMEOUT # If value is 0,"0","" or not passed then it set default to 300 seconds.
```
Comment (dbshah1212) [inline tap_exacttarget/client.py:93]:
```suggestion
            request_timeout = REQUEST_TIMEOUT # If value is 0, "0", "" or not passed then it sets default to 300 seconds.
```

Comment (savan-chovatiya) [inline tap_exacttarget/client.py:93]: @harshpatel4crest Can we remove this as request_timeout is already calculated above in this function itself?

File: tap_exacttarget/dao.py [hunk 4]
```diff
@@ -36,9 +38,31 @@ def load_schema(stream):
...
     return schema
 
+# boolean function to check if the error is 'timeout' error or not
+def is_timeout_error():
+    """
+        This function checks whether the URLError contains 'timed out' substring and return boolean
+        values accordingly, to decide whether to backoff or not.
+    """
+    def gen_fn(exc):
+        if str(exc).__contains__('timed out'):
```
Comment (dbshah1212) [inline tap_exacttarget/dao.py:48]: You simply check this by `if 'timed out' in str(exc)`

File: tap_exacttarget/dao.py [hunk 5]
```diff
@@ -36,9 +38,31 @@ def load_schema(stream):
...
 # decorator for retrying on error
 def exacttarget_error_handling(fnc):
-    @backoff.on_exception(backoff.expo, (socket.timeout, ConnectionError), max_tries=5, factor=2)
+    @backoff.on_exception(backoff.expo,
+                          urllib.error.URLError, # backoff 'timeout' error for SOAP API
+                          giveup=is_timeout_error(),
+                          max_tries=5,
+                          factor=2)
+    @backoff.on_exception(backoff.expo,
+                          (socket.timeout, ConnectionError, requests.Timeout),
```
Comment (dbshah1212) [inline tap_exacttarget/dao.py:63]: Do we need request.Timeout, socket.timeout, and URlError?

Comment (dbshah1212) [reply tap_exacttarget/dao.py:63]: Where is the import for ConnectionError?

Comment (harshpatel4crest) [reply tap_exacttarget/dao.py:63]: The socket.timeout was added as part of TDL-14621: Add retry logic to requests and TDL-14622: Retry ConnectionResetErrors.

This ConnectionError is the error regarding the backoff we added for ConnectionResetErrors

File: tap_exacttarget/client.py [hunk 6]
```diff
@@ -47,12 +50,20 @@ def get_auth_stub(config):
         params['soapendpoint'] = ('https://{}.soap.marketingcloudapis.com/Service.asmx'
                                   .format(config['tenant_subdomain']))
 
+    # Set request timeout with config param `request_timeout`.
+    config_request_timeout = config.get('request_timeout')
+    if config_request_timeout and float(config_request_timeout):
+        request_timeout = float(config_request_timeout)
+    else:
+        request_timeout = REQUEST_TIMEOUT # If value is 0, "0", "" or not passed then it sets default to 300 seconds.
```
Comment (namrata270998) [inline tap_exacttarget/client.py:58]: @harshpatel4crest Place this comment above if

File: tests/unittests/test_backoff.py [hunk 7]
```diff
@@ -808,3 +812,401 @@ def test_no_socket_timeout_error_occurred__subscriber(self, mocked_write_records
         # verify if 'tap_exacttarget.dao.DataAccessObject.write_records_with_transform' was called
         # once as there is only one record
         self.assertEquals(mocked_write_records.call_count, 1)
+
+@mock.patch("time.sleep")
+class TestTimeoutError(unittest.TestCase):
```
Comment (namrata270998) [inline tests/unittests/test_backoff.py:817]: @harshpatel4crest  Can we add a testcase when URLError occurs with string that **does not contain** timed out?

File: tap_exacttarget/client.py (hunk 3 above)
Comment (harshpatel4crest) [reply tap_exacttarget/client.py]: Done, added single code change to set timeout value

Comment (harshpatel4crest) [reply tap_exacttarget/client.py]: Done

File: tap_exacttarget/client.py (hunk 2 above)
Comment (harshpatel4crest) [reply tap_exacttarget/client.py]: Done

File: tap_exacttarget/client.py (hunk 1 above)
Comment (harshpatel4crest) [reply tap_exacttarget/client.py]: Done

File: tests/unittests/test_backoff.py (hunk 7 above)
Comment (harshpatel4crest) [reply tests/unittests/test_backoff.py:817]: Added test case.

File: tap_exacttarget/client.py (hunk 6 above)
Comment (harshpatel4crest) [reply tap_exacttarget/client.py]: Done

**Review by namrata270998 — APPROVED** [review]

File: tests/unittests/test_backoff.py [hunk 8]
```diff
@@ -808,3 +812,416 @@ def test_no_socket_timeout_error_occurred__subscriber(self, mocked_write_records
...
+        # make the object of 'SendDataAccessObject'
+        obj = sends.SendDataAccessObject({'start_date': '2021-01-01T00:00:00Z'}, {}, None, {})
+        # call sync
+        obj.sync_data()
+        # verify if 'tap_exacttarget.dao.DataAccessObject.write_records_with_transform' was called
+        # once as there is only one record
+        self.assertEquals(mocked_write_records.call_count, 1)
+
+    @mock.patch("FuelSDK.rest.ET_GetSupport.get")
+    def test_timeout_error_occurred__subscriber(self, mocked_get, mocked_sleep):
```
Comment (dbshah1212) [inline tests/unittests/test_backoff.py:1183]: please add function comments for all tests

Comment (harshpatel4crest) [reply tests/unittests/test_backoff.py:1393]: Added

**Review by dbshah1212 — APPROVED** [review]

File: tap_exacttarget/dao.py [hunk 9]
```diff
@@ -36,9 +38,31 @@ def load_schema(stream):
 
     return schema
 
+# boolean function to check if the error is 'timeout' error or not
+def is_timeout_error():
+    """
+        This function checks whether the URLError contains 'timed out' substring and return boolean
+        values accordingly, to decide whether to backoff or not.
+    """
+    def gen_fn(exc):
```
Comment (karanpanchal-crest) [inline tap_exacttarget/dao.py:47]: @harshpatel4crest Check if having on `gen_fn()` works for timeout

File: tests/unittests/test_backoff.py [hunk 10]
```diff
@@ -287,21 +337,28 @@ def test_connection_reset_error_occurred__list_subscriber(self, mocked_get, mock
...
+            Test case to verify that we backoff for 5 times when 'ConnectionResetError' error occurs
+        """
         # mock 'get' and raise error
         mocked_get.side_effect = socket.error(104, 'Connection reset by peer')
         # make the object of 'ListSubscriberDataAccessObject'
         obj = list_subscribers.ListSubscriberDataAccessObject({"start_date": "2020-01-01T00:00:00Z"}, {}, None, {})
         try:
             # call function
-            obj._get_all_subscribers_list()
+            obj.sync_data()
```
Comment (karanpanchal-crest) [inline tests/unittests/test_backoff.py:351]: Can you explain this code change?

**Review by karanpanchal-crest — CHANGES_REQUESTED** [review]

File: tests/unittests/test_backoff.py (hunk 10 above)
Comment (harshpatel4crest) [reply tests/unittests/test_backoff.py:351]: We have added backoff on `sync_data`, as the function `_get_all_subscribers_list` will be called from that function and if we add backoff on the **_get_all_subscribers_list**, there are chances of the Tap backing off 25 times.

File: tap_exacttarget/dao.py (hunk 9 above)
Comment (harshpatel4crest) [reply tap_exacttarget/dao.py]: Added a single function for giveup condition.

**Review by karanpanchal-crest — APPROVED** [review]

File: tests/unittests/test_error_messages.py [hunk 11]
```diff
@@ -28,7 +28,7 @@ def test_error_1(self, mocked_post_request):
             "client_secret": "client_secret_123",
             "tenant_subdomain": "",
             "start_date": "2019-01-01T00:00:00Z",
-            "request_timeout": "900",
+            "request_timeout": "300",
```
Comment (KrisPersonal) [inline tests/unittests/test_error_messages.py:31]: Why the timeout was changed from 900 to 300? There must be a reason it was set to 900.

Comment (harshpatel4crest) [reply tests/unittests/test_error_messages.py:31]: @KrisPersonal To maintain consistency among all the Tap, we have asked before implementing.
Please find the reference link: https://talend.slack.com/archives/C01TGC4QAUE/p1635252663017600

**Review by kspeer825 — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Pull request merged**